addons/account/models/partner.py

```diff
@@ -205,46 +205,42 @@ def write(self, vals):
                 vals['zip_from'], vals['zip_to'] = self._convert_zip_values(zip_from or rec.zip_from, zip_to or rec.zip_to)
         return super(AccountFiscalPosition, self).write(vals)
 
-    def _get_fpos_ranking_functions(self, partner):
-        """Get comparison functions to rank fiscal positions.
-
-        All functions are applied to the fiscal position and return a value.
-        These values are taken in order to build a tuple that will be the comparator
-        value between fiscal positions.
-
-        If the value returned by one of the function is falsy, the fiscal position is
-        filtered out and not even considered for the ranking.
-
-        :param partner: the partner to consider for the ranking of the fiscal positions
-        :type partner: :class:`res.partner`
-        :return: a list of tuples with a name and the function to apply. The name is only
-            used to facilitate extending the comparators.
-        :rtype: list[tuple[str, function]
+    def _get_first_matching_fpos(self, partner):
+        sorted_fpos = self.sorted(key=lambda f: (-len(f.company_id.parent_ids), f.sequence))  # company specific first, then sequence
+        for fpos in sorted_fpos:
+            if all(fn(fpos) for fn in self._get_fpos_validation_functions(partner)):
+                return fpos
+        return self.env['account.fiscal.position']
+
+    def _get_fpos_validation_functions(self, partner):
+        """ Returns a list of functions to validate fiscal positions against a partner.
         """
         return [
-            ('vat_required', lambda fpos: (
```

Comment:
This has a label on purpose so that we can update the function easily

Reply:
Yes, I didn't read the commit properly. But I don't understand the purpose of the PR; neither the commit message nor the task description clarifies it...

Reply:
As I understand, the ranking functions make it hard to determine which fiscal position will be chosen... I think it's currently strictest first.

IIUC this gives strange results - for example a sale in BE to a belgian with no vat will get the EU B2C fiscal position - this is functionally ok, but confusing for users.